python/cudf/cudf/core/dataframe.py

```diff
@@ -3795,6 +3796,309 @@ def replace(
+
+        The `memory_usage` parameter allows deep introspection mode, specially
+        useful for big DataFrames and fine-tune memory optimization:
+
+        >>> import numpy as np
+        >>> random_strings_array = np.random.choice(['a', 'b', 'c'], 10 ** 6)
+        >>> df = cudf.DataFrame({
+        ...     'column_1': np.random.choice(['a', 'b', 'c'], 10 ** 6),
+        ...     'column_2': np.random.choice(['a', 'b', 'c'], 10 ** 6),
+        ...     'column_3': np.random.choice(['a', 'b', 'c'], 10 ** 6)
+        ... })
+        >>> df.info(memory_usage='deep')
+        <class 'cudf.core.dataframe.DataFrame'>
+        RangeIndex: 1000000 entries, 0 to 999999
+        Data columns (total 3 columns):
+         #   Column    Non-Null Count    Dtype
+        ---  ------    --------------    -----
+         0   column_1  1000000 non-null  object
+         1   column_2  1000000 non-null  object
+         2   column_3  1000000 non-null  object
+        dtypes: object(3)
+        memory usage: 14.3 MB
+        """
+        if buf is None:
+            buf = sys.stdout
+
+        lines = [str(type(self))]
+
+        index_name = type(self._index).__name__
+        if len(self._index) > 0:
+            entries_summary = f", {self._index[0]} to {self._index[-1]}"
+        else:
+            entries_summary = ""
+        index_summary = (
+            f"{index_name}: {len(self._index)} entries{entries_summary}"
+        )
+        lines.append(index_summary)
+
+        if len(self.columns) == 0:
+            lines.append(f"Empty {type(self).__name__}")
+            cudf.utils.ioutils.buffer_write_lines(buf, lines)
+            return
+
+        cols = self.columns
+        col_count = len(self.columns)
+
+        if max_cols is None:
+            max_cols = pd.options.display.max_info_columns
+
+        max_rows = pd.options.display.max_info_rows
+
+        if null_counts is None:
+            show_counts = (col_count <= max_cols) and (len(self) < max_rows)
+        else:
+            show_counts = null_counts
+
+        exceeds_info_cols = col_count > max_cols
+
+        def _put_str(s, space):
+            return str(s)[:space].ljust(space)
+
+        def _verbose_repr():
+            from pandas.io.formats.printing import pprint_thing
+
+            lines.append(f"Data columns (total {len(self.columns)} columns):")
+
+            id_head = " # "
+            column_head = "Column"
+            col_space = 2
+
+            max_col = max(len(pprint_thing(k)) for k in cols)
+            len_column = len(pprint_thing(column_head))
+            space = max(max_col, len_column) + col_space
+
+            max_id = len(pprint_thing(col_count))
+            len_id = len(pprint_thing(id_head))
+            space_num = max(max_id, len_id) + col_space
+            counts = None
+
+            header = _put_str(id_head, space_num) + _put_str(
+                column_head, space
+            )
+            if show_counts:
+                counts = self.count()
```

Comment:
Should we explicitly copy this back to host since we're going to be doing fine grained accesses on it serially in Python?